### PR TITLE
Add translation statistics per language

### PR DESCRIPTION
**Pull request by kumakyoo42**
When I visited [sentences by language](https://tatoeba.org/de/stats/sentences_by_language) for the first time and clicked on a language, I expected some statistics for that language (you get a list of all sentences of that language instead). As this sort of statistics is not available I thought it might be a good idea to add it.

From my experience as a PR reviewer I know it's better to keep PRs small. Hence, this is only part of the work - I did not yet add what will happen when no language is given, or when the language given does not exist. I plan to add this in future PRs. (If you prefere larger all at once PRs, please tell me!)

As this change isn't complete, I didn't change the above mentioned links yet, so for testing you need to point your browser to .../stats/statistics_for_language/eng. 

There is one issue about translations, I couldn't solve yet. Maybe you can help me out here: The sentence above the table "There are 9 sentences in Deutsch, 7 of them have at least one translation." contains two numbers which would change the sentence when being the value 1. As there are two numbers, I can't use `__n()` here or at least I'm not aware of how to do this. Any ideas?

And one more thing: My editor automatically removes trailing spaces and I don't know, how to stop it from doing so. Hence there are a few unrelated minor changes in this PR. Hope this doesn't hurt.

### CODE REVIEW

File: src/Controller/StatsController.php [hunk 1]
```diff
@@ -55,6 +56,37 @@ function sentences_by_language() {
...
 
 
+    /**
+     * Show statistics for a specified language.
+     *
+     * @param mixed $lang Language for which statistics should be displayed.
+     *
+     * @return void
+     */
+    function statistics_for_language($lang = null)
```
Comment (jiru) [inline src/Controller/StatsController.php]: Please try to use an action name that makes a [nice URL](https://en.wikipedia.org/wiki/Clean_URL). The URL `http://localhost:8080/fr/stats/statistics_for_language/eng` includes both words stats and statistics, which is redundant. You could change the action name here, or alternatively reuse the existing `sentences_by_language` action, so it becomes `/stats/sentences_by_language/eng`.

File: src/Model/Table/SentencesTable.php [hunk 2]
```diff
@@ -1393,4 +1393,55 @@ public function languagesHavingSentences()
...
+                'conditions' => 'Sentences.id = T.sentence_id',
+            ])
+            ->select([
+                'code' => 'T.translation_lang',
+                'translations' => 'COUNT(*)',
+            ])
+            ->where(['Sentences.lang' => $lang])
+            ->group(['code'])
+            ->order(['translations' => 'DESC'])
+            ->toList();
```
Comment (jiru) [inline src/Model/Table/SentencesTable.php]: This request is too expensive to be executed just like that. To give you an idea, I tried to execute it for English on https://dev.tatoeba.org/, on which `sentences_translations` contains 15,437,391 records, and it took 33 seconds. By the way, https://tatoeba.org/ ’s `sentences_translations` table contains 27,688,900 records at the moment.

We can consider executing it as a cronjob, or caching the result of the query.

Also I think you don’t need to do any JOIN here. You can simply use `sentences_translations.sentence_lang` instead of joining with `sentences` to get `sentences.lang`.

File: src/Model/Table/SentencesTable.php [hunk 3]
```diff
@@ -1393,4 +1393,55 @@ public function languagesHavingSentences()
...
+                'table' => 'sentences_translations',
+                'alias' => 'T',
+                'type' => 'RIGHT',
+                'conditions' => 'Sentences.id = T.sentence_id',
+            ])
+            ->select([
+                'count' => 'COUNT(DISTINCT(T.sentence_id))',
+            ])
+            ->where(['Sentences.lang' => $lang])
+            ->first()->count;
```
Comment (jiru) [inline src/Model/Table/SentencesTable.php]: Same remarks as for `getTranslationStatistics()`.

In addition, consider using `firstOrFail()`. Alternatively, directly call `->count()`, which makes CakePHP wrap your query into a `COUNT()`.

File: src/Template/Stats/statistics_for_language.ctp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    ?>
+
+    <table class="language-stats">
+    <tr>
+        <th></th>
+        <th></th>
+        <th></th>
+        <?php /* @translators: table header text in "Statistics for language" page */ ?>
+        <th><?php echo __('Language'); ?></th>
+        <th><?php echo __('Translations'); ?></th>
```
Comment (jiru) [inline src/Template/Stats/statistics_for_language.ctp]: Consider adding a `@translators:` note above `__('Translations')`, too. The notes need to be on the line immediately preceding the `__()` call.

File: tests/TestCase/Model/Table/LanguagesTableTest.php [hunk 5]
```diff
@@ -74,6 +74,10 @@ function testGetTotalSentencesNumber() {
         $expected = 56;
         $n = $this->Languages->getTotalSentencesNumber();
         $this->assertEquals($expected, $n);
+        
+        $expected = 21;
+        $n = $this->Languages->getTotalSentencesNumber('eng');
+        $this->assertEquals($expected, $n);
```
Comment (jiru) [inline tests/TestCase/Model/Table/LanguagesTableTest.php]: Can you add a test case for a invalid language code?

File: webroot/css/stats/statistics_for_language.css [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+.language-stats img {
+    display: block;
+}
+
+.language-stats .status-warning {
+    color: #AA0000;
+    margin-left: 10px;
+    float: right;
+}
```
Comment (jiru) [inline webroot/css/stats/statistics_for_language.css]: I don’t think the `status-warning` class appears anywhere in the statistics_for_language view.

Comment (kumakyoo42) [reply webroot/css/stats/statistics_for_language.css]: I wasn't sure if this might creep in from other code I do not know about, so I kept it (I copy-paste from `sentences_by_language.css`). I'll remove it.

**Comment by kumakyoo42** [comment]
I think I addressed every one of your comments now. 

I copied the database to my local version to did some speed tests. Unfortunately querying the English corpus is still too slow (20 seconds on my computer). I learned that InnoDB cannot index `count(*)` and therefore the query can't be speed up. So, in my opinion, we have to resort to something else, like cronjob or caching as you mentioned. (Alternativley we could add a translations table similar to the languages table, but keeping this up to date would not be worth it, I think.)

Suggestion: I'll fill in the "Not yet implemented" parts first, before we dicide on how to continue.

I also want to do the cherry picking for removing the unrelated trailing-whitespace lines from this PR. I didn't do that yet, but I'd like to learn how this works.

File: src/Controller/StatsController.php [hunk 7]
```diff
@@ -55,6 +56,35 @@ function sentences_by_language() {
...
+     * @return void
+     */
+    function translations_by_language($lang = null)
+    {
+        if (!array_key_exists($lang, LanguagesLib::languagesInTatoeba())) {
+            $lang = null;
+        }
+
+        $this->set('lang', $lang);
+        if ($lang === null) return;
```
Comment (jiru) [inline src/Controller/StatsController.php]: I think these two ifs could be simplified:
1. Move the return statement into the first if:
```php
if (!array_key_exists($lang, LanguagesLib::languagesInTatoeba())) {
    return;
}
```
2. In the view, replace `$lang === null` with `isset($lang)`

File: tests/TestCase/Model/Table/LanguagesTableTest.php [hunk 8]
```diff
@@ -74,6 +74,13 @@ function testGetTotalSentencesNumber() {
         $expected = 56;
         $n = $this->Languages->getTotalSentencesNumber();
         $this->assertEquals($expected, $n);
+
+        $expected = 21;
+        $n = $this->Languages->getTotalSentencesNumber('eng');
+        $this->assertEquals($expected, $n);
+
+        $n = $this->Languages->getTotalSentencesNumber('und');
+        $this->assertNull($n);
```
Comment (jiru) [inline tests/TestCase/Model/Table/LanguagesTableTest.php:85]: Could you also test `$this->Languages->getTotalSentencesNumber(null)`? Alternatively, add a string type constraint on the `$lang` parameter of `getTotalSentencesNumber()`.

**Comment by jiru** [comment]
Here are some stats of running your SQL requests on tatoeba.org:

* getTranslationStatistics (without join):
  * eng: 4 to 6s
  * epo: 1 to 1.5s
  * jpn: 0.3 to 0.5s

* getTotalTranslations (with join, which is currently faster than without):
  * eng: 12 to 15s
  * epo: 4 to 6s
  * jpn: 1.2 to 1.8s

I think this is reasonable if we compute the stats for all languages with a daily cronjob. You can create a simple Cake shell that calculates all the stats and write them to a cache store (see CakePHP Cache class) and retrieve this data on page load.

There is one issue with this approach though: we clear all cache when updating the website, so after a site update, the stats would be unavailable until the next cronjob execution. I think this can be solved by setting up a special kind of persistent cache (in `config/app.php`) that survives website updates and server reboot.

> Suggestion: I'll fill in the "Not yet implemented" parts first, before we dicide on how to continue.

Sure.

> I also want to do the cherry picking for removing the unrelated trailing-whitespace lines from this PR. I didn't do that yet, but I'd like to learn how this works.

* To stage individual blocks to your diff, before commiting: `git add --patch`
* To unstage  individual blocks from your diff: `git restore --staged --patch`

File: src/Model/Table/SentencesTable.php [hunk 9]
```diff
@@ -1393,4 +1393,49 @@ public function languagesHavingSentences()
...
+    public function getTranslationStatistics($lang)
+    {
+        $results = $this->Links->query()
+            ->select([
+                'code' => 'translation_lang',
+                'translations' => 'COUNT(*)',
+            ])
+            ->where(['sentence_lang' => $lang])
+            ->group(['code'])
+            ->order(['translations' => 'DESC'])
```
Comment (jiru) [inline src/Model/Table/SentencesTable.php]: When two languages have the same number of translations, their order is undefined so their ranking could be swapped randomly between different runs. You could add a second ordering criterion to make the ordering consistent.

**Comment by jiru** [comment]
After removing the JOIN clause, on tatoeba.org and dev.tatoeba.org the stats are different, some of the numbers got slightly bigger.

I guess this must be caused by dangling links (records in `sentences_translations` that do not refer to existing sentences), or mismatches between `sentence.lang` and `sentences_translations.sentence_lang`.

The different is not so big.

Maybe this can be considered as a different issue we may separately solve.

**Comment by kumakyoo42** [comment]
- I added a global table (when there is no language given) showing for every language the number of sentences with at least one translation.
- I added language selection in the side bar.
- I added a link from the homepage statistics to this page. (I changed my mind here, changing the links on the language statistics page seems not apropriate to me any more.)
- I replaced the links in the table to point to an (advanced) search showing all sentences which are counted in this row. I couldn't test that though, because search doesn't work locally on my computer. Don't know, what I have to do here.
- I addressed all your concerns above.

If that's all fine, the only thing, that misses now (from my point of view), is the use of a cronjob and cached data.

**Comment by jiru** [comment]
It looks very good, thank you.

> * I added a link from the homepage statistics to this page. (I changed my mind here, changing the links on the language statistics page seems not apropriate to me any more.)

It think it is reasonable to add the link on the homepage, but we now have a weird blank space on the left block:

image

I am not sure what’s the best approach. We could add more text to the left part to reduce the void, or move each stats link into the text, for example "25,881 [sentences with translations](http://localhost:8080/en/stats/translations_by_language)".

> search doesn't work locally on my computer

It works for me. You can check `/var/manticore/log/searchd.log`.

File: src/Controller/StatsController.php [hunk 10]
```diff
@@ -55,6 +56,34 @@ function sentences_by_language() {
...
+     * Show translation statistics for a specified language.
+     *
+     * @param mixed $lang Language for which statistics should be displayed.
+     *
+     * @return void
+     */
+    function translations_by_language($lang = null)
+    {
+        if (!array_key_exists($lang, LanguagesLib::languagesInTatoeba())) {
+            $lang = null;
```
Comment (jiru) [inline src/Controller/StatsController.php]: If the language is provided but invalid, I think we should show a 404 error.

With the current implementation, opening http://localhost:8080/en/stats/translations_by_language/foobar shows "Translation statistics for all languages".

To show a 404 error, simply throw a Cake\Http\Exception\NotFoundException.

File: src/Template/Stats/translations_by_language.ctp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+<div id="main_content">
+<div class="section md-whiteframe-1dp">
+    <h2><?= $title; ?></h2>
+
+    <p>
+    <?php
+        if (empty($lang)) {
+            echo '<p>'.format(
+                __n('There is one sentence in the database.',
+                'There are {total}&nbsp;sentences in the database.',
```
Comment (jiru) [inline src/Template/Stats/translations_by_language.ctp]: I’d like to suggest using the word "corpus" instead of "database". I feel like "database" is unnecessarily technical.

File: src/Template/Stats/translations_by_language.ctp [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+
+        $langName = $this->Languages->codeToNameAlone($langCode);
+        if (empty($langCode)) {
+            $langCode = 'unknown';
+        }
+        $languageLink = $this->Html->link(
+            $langName,
+            array(
+                "controller" => "sentences",
+                "action" => "search?from=".htmlspecialchars($from)
```
Comment (jiru) [inline src/Template/Stats/translations_by_language.ctp]: * You can have `link()` build the query string with:
  ```php
  [
    'controller' => 'sentences',
    'action' => 'search',
    '?' => [
      'from' => $from,
      /* etc. */
    ]
  ]
  ```
* You should add `to=$from` in addition to `trans_to=$from`. `trans_to=` is to limit to sentences having translations, while `to=` is to limit what translations are showed.

File: src/Template/Stats/translations_by_language.ctp [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+            )
+        );
+
+        echo '<tr>';
+
+        echo $this->Html->tag('td', $this->Number->format($rank));
+        echo $this->Html->tag('td', $languageIcon);
+        echo $this->Html->tag('td', $langCode);
+        echo $this->Html->tag('td', $languageLink);
+        echo $this->Html->tag('td', $numSentencesDiv, array('class' => 'num-sentences'));
```
Comment (jiru) [inline src/Template/Stats/translations_by_language.ctp:150]: It looks like this `num-sentences` class is not used. It is used in `sentences_by_language.css`, which is not loaded here.

File: src/Template/Stats/translations_by_language.ctp [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+            $from = $langCode;
+            $to = '';
+        } else {
+            $from = $lang;
+            $to = $langCode;
+        }
+
+        $langName = $this->Languages->codeToNameAlone($langCode);
+        if (empty($langCode)) {
+            $langCode = 'unknown';
```
Comment (jiru) [inline src/Template/Stats/translations_by_language.ctp:122]: Clicking on the "unknown language" link in the table brings to a search page that does not show sentences in "unknown language" or having translations in "unknown language". Currently, the search engine does not index sentences in "unknown" language, so you cannot display such sentences. So I think you should not show any link in that case.

**Comment by kumakyoo42** [comment]
> It think it is reasonable to add the link on the homepage, but we now have a weird blank space on the left block:

Sorry, I was only aware of that block being used in the right side column.

Maybe we could add a link to https://tatoeba.org/de/donate at the left box?

**Comment by kumakyoo42** [comment]
> It works for me. You can check /var/manticore/log/searchd.log.

See issue #3258 why it didn't work for me.

**Comment by kumakyoo42** [comment]
`createFilterByLangMod` lists the languages of the profile of the user, but it should list the languages in the table. Do you know an easy way to get the correct list displayed?

**Comment by jiru** [comment]
> Maybe we could add a link to https://tatoeba.org/de/donate at the left box?

Good idea. I can imagine a text saying something like "If you like what we do, consider making a [donation](https://tatoeba.org/donate)."

> `createFilterByLangMod` lists the languages of the profile of the user, but it should list the languages in the table. Do you know an easy way to get the correct list displayed?

You could allow the list of languages to be passed as a new parameter to `createFilterByLangMod`. Just to clarify, the profile languages show up in the dropdown when opening it, but any language can be selected by typing something.

**Comment by kumakyoo42** [comment]
> Just to clarify, the profile languages show up in the dropdown when opening it, but any language can be selected by typing something.

Ah, I didn't know that. So no change is needed here.

**Comment by kumakyoo42** [comment]
I still need to implement the caching stuff. But all else should be fine now.

**Comment by kumakyoo42** [comment]
I meanwhile read a lot in the CakePHP documentation about shell and caching, but I still have no clue where to start at all with a cached cron job.

In my opinion CakePHP complicates things a lot and I realize that I start loosing interest in contributing at all.

Sorry, but I will not continue here. If someone want's to take over, feel free to do so.
